lib/ukstore: Introduce dynamic store API

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This series adds dynamic entry support to uk_store. Libraries can dynamically register uk_store objects at runtime, each one of which is associated with one or more entries. Similarly to static entries, each dynamic entry is associated with a getter and / or a setter.

Consumers are notified for the lifetime of objects via events.

GitHub-Supersedes: #460
GitHub-Depends-On: #938

michpappas committed:
lib/ukstore: Remove cookie from static entries

Remove cookie from static uk_store entries. This is in anticipation of
dynamic entries that provide a cookie at the object level.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukalloc: Remove deprecated cookie parameter from static entry definitions

Update calls to UK_STORE_STATIC_ENTRIES to remove the deprecated
cookie parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukalloc: Add entry IDs and update static entry definitions

Introduce header for definitions related to uk_store. Add
entry IDs for stats tracked by uk_alloc, and update static
entry definitions to pass entry IDs.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukstore: Lookup entries by id

Introduce an id property to uk_store entries and replace name with id
on lookup operations, to avoid costly string comparison.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

craciunoiuc committed:
lib/ukstore: Introduce dynamic API

This commit adds dynamic entry support to uk_store. Libraries can
dynamically register uk_store objects at runtime, each one of which
is associated with one or more entries. Similarly to static entries,
each dynamic entry is associated with a getter and / or a setter.

Consumers are notified for the lifetime of objects via events.

Checkpatch-Ignore: SPACING LONG_LINE
GitHub-Closes: #539
Co-authored-by: Michalis Pappas <michalis@unikraft.io>
Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukstore: Add README.md

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

craciunoiuc commented on lib/ukalloc/include/uk/alloc_store.h:
```diff
@@ -0,0 +1,19 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* FIXME */
```
Missed comment

craciunoiuc commented on lib/ukstore/README.md:
```diff
@@ -0,0 +1,175 @@
+# uk_store
+
+uk_store provides Unikraft libraries with a generic API of getters and setters. The API defines a producer / consumer model that implements pull-oriented getters and push-oriented setters. Under that model, producer libraries define and export a set of entries, each one of which consists of a single value. Consumer libraries retrieve and / or update entries using their associated getters and sett ...
```
One thing I like to do in `md` files is to break paragraphs in one per line. It makes reviewing easier and future additions.

Change only if you want ;)

craciunoiuc commented on lib/ukstore/README.md:
````diff
@@ -0,0 +1,175 @@
+### Example
+
+#### Device initialization: Producer
+```
````
\`\`\`c

craciunoiuc commented on lib/ukstore/README.md:
```diff
@@ -0,0 +1,175 @@
+The static API allows libraries to define a flat list of entries, ie without parent object. The hierarchy is then defined as `library/entry[value]`. Static entries are at compile-time via the `UK_STORE_STATIC_ENTRY` macro.
+
+The static API is a reduced version of the dynamic API as:
+* There is no requirement for dynamic memory allocation.
```
Codacy warnings, you need 2 spaces before `*`

craciunoiuc reviewed (changes requested):
Some small comments from my side, it looks good 🙏 


I hope you also modified the code to be in accordance with the new coding conventions 😅

craciunoiuc commented:
I saw also some errors in checkpatch, if those are actually ok, add the `Checkpatch-Ignore: ...` in the message

craciunoiuc commented:
In the PR text also add `Closes` for automatic closing when merged

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests

craciunoiuc commented:
When @skuenzer is happy I'll also add my tag and we can merge

michpappas commented:
Force pushed to update author / SOB and clearly mark dependencies.

craciunoiuc commented on lib/ukstore/README.md:
```diff
@@ -0,0 +1,189 @@
+  * uk_store events are not used.
+  * Consumers access entries directly, without the need to call `uk_store_acquire_object()`. Calling `uk_store_get_entry()` requires passing `UK_STORE_OBJECT_ID_NONE` to the `object_id` parameter.
+
+The static API is useful inthe cases where a library exports a fixed set of entries that do not depend on individual instances.
```
"The static API is useful `in the` cases"

michpappas replied:
Done

craciunoiuc commented:
@michpappas @skuenzer is the dependency for this merged? Can it be rebased?

adinamariav reviewed (commented):
Reviewed-by: Adina-Maria Vaman <adinamariav22@gmail.com>

michpappas commented:
> @michpappas @skuenzer is the dependency for this merged? Can it be rebased?

Rebased to `staging`.

craciunoiuc commented:
@michpappas your user is not recognised by github, did you add your unikraft.io email to github?

![image](https://github.com/unikraft/unikraft/assets/46280801/3b174325-33d0-4585-a6bb-71759c7d16c6)

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
-typedef int (*uk_store_get_u64_func_t)(void *, __u64 *);
-typedef int (*uk_store_get_uptr_func_t)(void *, __uptr *);
-typedef int (*uk_store_get_charp_func_t)(void *, char **);
+typedef int (*_uk_store_get_s8_func_t)(void *, __s8 *);
```
Is there a reason for the leading underscore for a datatype definition? It is already namespaced with `uk_`

michpappas replied:
I don't think so, will remove.

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
+	__u64 id;
+
+	/* The entry name */
+	char *name;
```
Probably we should declare this as `const char *`?

michpappas replied:
Same comment as in https://github.com/unikraft/unikraft/pull/939#discussion_r1296271248

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
+	int flags;
+
+	/* Entry getter/setter type */
+	enum uk_store_entry_type type;
```
Would it make sense to move the type to the op struct? Then, it would be closely related to the actual function pointers. Alternatively we could also embed the declaration of `ops` in here.

michpappas replied:
~~Agreed, will move the ops into the main object.~~ In fact, this was introduced by me during refactoring in 97eedfb247ae1f00b15c941fdef3b9f3d6ef3118. Will revert.

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
+	__u64 id;
+
+	/* The object name */
+	char *name;
```
`const char *`?

michpappas replied:
These cannot be `const char *` without changing the implementation because `uk_store_allocate_object()` and `_UK_STORE_DYNAMIC_CREATE_TYPED` copy the `const char *` name passed by the caller into a locally allocated buffer. Not sure why it was done like that, so I assumed there was good reason to do so (although I still don't see it).

skuenzer replied:
This makes sense actually. An API user likely generates object names from a template and you do not want to deal with keeping track of the string buffer. :+1:

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
+ * @param library_id owner library id as returned by uklibid
+ * @param object_id  object id
+ */
+void uk_store_release_object(__u16 library_id, __u64 object_id);
```
Would `void uk_store_release_object(struct uk_store_object *obj);` work as well? I remember we had a longer discussion about this but I don't remember the result.

michpappas replied:
From a look at the client side, I see no reason why we can't do that. Yet this needs to be changed with clear mind as it affects how we lock the refcounts. I will look at it tomorrow morning.

skuenzer replied:
I agree, sounds good.

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
+		.id = _id,					\
+		.name = STRINGIFY(_name),			\
+		.type = UK_STORE_ENTRY_TYPE(_type),		\
+		.ops.get.u64 = _get,				\
```
`_type` instead of `_u64`

michpappas replied:
Ok

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
+		_uk_store_set_charp_func_t charp;
+	} set;
+
+	_uk_store_entry_freed_func_t cleanup;
```
Do we need this hook? Shall we remove it?

michpappas replied:
We don't use it so far, will remove

skuenzer replied:
:+1:

skuenzer commented on lib/ukstore/include/uk/store.h:
```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
 /* Library-specific information generated at compile-time */
 #include <uk/bits/store_libs.h>
 
+struct uk_store_ops {
```
Embed or move type into this struct

michpappas commented:
@skuenzer I addressed all review comments except from https://github.com/unikraft/unikraft/pull/939#discussion_r1296018109 which I will look at tomorrow morning. I marked the rest as resolved.

michpappas commented:
@skuenzer all changes are in now, including an updated `README.md`.

skuenzer commented:
@michpappas : I just rebased this PR to latest `staging` and removed the commit "lib/ukstore: Bring function definitions into the private scope", as we discussed.

skuenzer reviewed (approved):
Thanks a lot for all your work @craciunoiuc and @michpappas 

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
